### ISSUE TITLE
'[ -t 1 ]' and 'exec >...' breakage  in ${ shared; } command substitutions

### ISSUE THREAD

**Issue by McDutchie** [ksh93/ksh#24]
Commit cafe33f048680e9b63e28ed0013c812c13b22863 (see there for links/documentation) fixed `test -t 1` in command substitutions by implementing a workaround: fork the subshell.

A real fix would be preferable, because `test -t 1` and `[ -t 1 ]` and `[[ -t 1 ]]` continue to be wrong for "shared" command substitutions of the form `var=${ list; }`:

```sh
$ v=${ [ -t 1 ] && echo whoops; }; echo $v
whoops
```

The forking workaround is not possible there, as forking would defeat the "shared" aspect of the shared command substitution.

**Comment by JohnoKing** [comment]
I have done some debugging, and it appears that `[ -t 1 ]` is broken in subshares and virtual subshells because `stdout` _is_ registered as a terminal:
https://github.com/ksh93/ksh/blob/876da711c0e6f7f73f5f8f232ebfdb77a09dc57b/src/cmd/ksh93/bltins/test.c#L317

If you replace all instances of `tty_check` in test.c with `isatty`, the problem still occurs. This bug is likely caused by the way ksh handles `stdout` in virtual subshells.

**Comment by McDutchie** [comment]
Yes, it does seem to explicitly register stdout as a terminal in non-forked command substitutions, because the bug still occurs if you make sure the surrounding standard output is *not* a terminal:

```sh
$ ksh -c '{ v=${ [ -t 1 ] && echo whoops; } >/dev/null; }; echo $v'
whoops
```

How silly. Because, by definition and by default, stdout is *not* on a terminal in any form of command substitution. The whole point of it is to capture command output.

So it seems the next step is to find the point in the code where it does that, change that to register it as *not* a terminal, and then see what regression tests break.

**Comment by McDutchie** [comment]
Ooh, here's another nice one:

```sh
$ ksh -c 'v=${ exec >/dev/tty; echo whoops }'
whoops
whoops
whoops
whoops
[etc. etc. etc. ad infinitum]
```

Infinite loop!

Identical behaviour in ksh2020.

**Comment by JohnoKing** [comment]
The loop caused by running `ksh -c 'v=${ exec >/dev/tty; echo whoops }'` was introduced in ksh93t+ 2009-05-01. I can't reproduce it in ksh93t+ 2008-12-10 (which the [multishell repo](https://github.com/multishell/ksh93/tree/ms_2008-12-12) calls ksh93t+ 2008-12-12). The `test -t 1` bug is much older though, as I can reproduce it in versions going back to at least ksh93n- 1993-12-28.

**Comment by McDutchie** [comment]
This bug (with `test -t 1`/`[ -t 1 ]` in subshares) was fixed in 55f0f8ce52ec39a797a745031a60816f998ce01a.

I'll open another issue for the infinite loop.
